File: SenLib/Pkg.cs
using HyoutaPluginBase;
using HyoutaPluginBase.FileContainer;
using HyoutaUtils;
using HyoutaUtils.Streams;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SenLib {
	public class Pkg : HyoutaPluginBase.FileContainer.IContainer {
		public uint Unknown; // seems like a unique index or something, at least in CS1/2?
		public List<PkgFile> Files;

		public Pkg(DuplicatableStream stream, EndianUtils.Endianness e = EndianUtils.Endianness.LittleEndian) {
			var s = stream.Duplicate();
			s.Position = 0;
			Unknown = s.ReadUInt32(e);
			uint fileCount = s.ReadUInt32(e);

			Files = new List<PkgFile>((int)fileCount);
			for (long i = 0; i < fileCount; ++i) {
				Files.Add(new PkgFile(s, e));
			}
		}

		public bool IsFile => false;

		public bool IsContainer => true;

		public IFile AsFile => null;

		public IContainer AsContainer => this;

		public void Dispose() {
			foreach (PkgFile f in Files) {
				f.Dispose();
			}
		}

		public INode GetChildByIndex(long index) {
			return Files[(int)index];
		}

		public INode GetChildByName(string name) {
			foreach (PkgFile f in Files) {
				if (f != null && f.Filename == name) {
					return f;
				}
			}
			return null;
		}

		public IEnumerable<string> GetChildNames() {
			foreach (PkgFile f in Files) {
				if (f != null && f.Filename != null) {
					yield return f.Filename;
				}
			}
		}

		public void WriteToStream(Stream s, EndianUtils.Endianness e) {
			long pos = s.Position;
			s.WriteUInt32(Unknown, e);
			s.WriteUInt32((uint)Files.Count, e);
			for (int i = 0; i < Files.Count; ++i) {
				var f = Files[i];
				s.WriteUTF8(f.Filename, 0x40, false);
				s.WriteUInt32(f.UncompressedSize, e);
				s.WriteUInt32(0, e); // compressed size, fill in later
				s.WriteUInt32(0, e); // file position, fill in later
				s.WriteUInt32(f.Flags, e);
			}

			for (int i = 0; i < Files.Count; ++i) {
				var f = Files[i];
				long currentFileStart = s.Position;
				using (DuplicatableStream fs = f.Data.Duplicate()) {
					uint p = (uint)(currentFileStart - pos);
					uint size = (uint)fs.Length;
					s.Position = pos + 8 + 0x50 * i + 0x44;
					s.WriteUInt32(size, e);
					s.WriteUInt32(p, e);
					s.Position = currentFileStart;
					StreamUtils.CopyStream(fs, s);
				}
			}
		}
	}

	public class PkgFile : HyoutaPluginBase.FileContainer.IFile {
		// individual file unpacking logic can be found at 0x41ad80 in CS2

		public string Filename;
		public uint UncompressedSize;
		public uint Flags;
		public DuplicatableStream Data;
		public EndianUtils.Endianness Endian;

		internal PkgFile(DuplicatableStream s, EndianUtils.Endianness e) {
			Endian = e;
			Filename = s.ReadSizedString(0x40, TextUtils.GameTextEncoding.UTF8).TrimNull();
			UncompressedSize = s.ReadUInt32(e); // 0x40
			uint compressedSize = s.ReadUInt32(e); // 0x44
			uint position = s.ReadUInt32(e); // 0x48
			Flags = s.ReadUInt32(e); // 0x4c
			Data = new PartialStream(s, position, compressedSize);
		}

		private static DuplicatableStream DecompressInternal(uint type, DuplicatableStream stream, long uncompressedSize, EndianUtils.Endianness e) {
			switch (type) {
				case 0: {
					// no compression
					return stream.Position != 0 ? new PartialStream(stream, stream.Position, stream.Length - stream.Position) : stream;
				}
				case 1: {
					// decompression algorithm derived from 0x41aa50 in CS2
					// very simple, can only copy verbatim bytes or reference already written bytes
					byte[] target = new byte[uncompressedSize];
					uint targetSize = stream.ReadUInt32(e); // seems to be unused
					uint sourceSize = stream.ReadUInt32(e);
					int backrefByte = stream.ReadInt32(e); // this is in fact read and compared as a 32-bit int
					uint targetPosition = 0;
					uint sourcePosition = 12;
					while (sourcePosition < sourceSize) {
						byte sourceByte = stream.ReadUInt8();
						++sourcePosition;
						if (sourceByte == backrefByte) {
							int backrefOffset = stream.ReadUInt8();
							++sourcePosition;
							if (backrefOffset == backrefByte) {
								target[targetPosition] = (byte)backrefByte;
								++targetPosition;
							} else {
								if (backrefByte < backrefOffset) {
									--backrefOffset;
								}
								uint backrefLength = stream.ReadUInt8();
								++sourcePosition;
								for (uint i = 0; i < backrefLength; ++i) {
									target[targetPosition + i] = target[targetPosition - backrefOffset + i];
								}
								targetPosition += backrefLength;
							}
						} else {
							target[targetPosition] = sourceByte;
							++targetPosition;
						}
					}
					if (targetSize != targetPosition) {
						Console.WriteLine("WARNING: Decompressed file is different filesize than expected. (type 1)");
					}
					return new DuplicatableByteArrayStream(target);
				}
				case 4: {
					// introduced in CS3 PC, I think, it's just LZ4
					byte[] compressed = stream.ReadBytes(stream.Length - stream.Position);
					byte[] decompressed = new byte[uncompressedSize];
					int result = lz4_sharp.lz4.LZ4_decompress_safe(compressed, 0, decompressed, 0, compressed.Length, decompressed.Length);
					if (result != decompressed.Length) {
						throw new Exception("LZ4 decompression failed.");
					}
					return new DuplicatableByteArrayStream(decompressed);
				}
				default:
					throw new Exception("Unsupported compression type '" + type + "'.");
			}
		}

		private DuplicatableStream Decompress() {
			bool hasChecksum = (Flags & 2u) != 0;
			using var stream = Data.Duplicate();
			stream.ReStart();
			uint checksum = hasChecksum ? stream.ReadUInt32(Endian) : 0;
			if (hasChecksum) {
				// verify checksum here? see 0x41ad40 in CS2, seems to be some kind of CRC
			}
			using var decompressed = DecompressInternal(Flags & ~2u, stream, UncompressedSize, Endian);
			if (decompressed.Length != UncompressedSize) {
				Console.WriteLine("WARNING: Decompressed file is different filesize than expected. (header)");
			}
			return decompressed.Duplicate();
		}

		public DuplicatableStream DataStream => Decompress();

		public bool IsFile => true;

		public bool IsContainer => false;

		public IFile AsFile => this;

		public IContainer AsContainer => null;

		public void Dispose() {
			if (Data != null) {
				Data.Dispose();
				Data = null;
			}
		}
	}
}

File: native/sen/pkg.cpp
#include "pkg.h"

#include "util/endian.h"
#include "util/memread.h"
#include "util/memwrite.h"

namespace SenLib {
PkgHeader::PkgHeader() = default;
PkgHeader::PkgHeader(PkgHeader&& other) = default;
PkgHeader& PkgHeader::operator=(PkgHeader&& other) = default;
PkgHeader::~PkgHeader() = default;

bool ReadPkgFromMemory(PkgHeader& pkg,
                       const char* buffer,
                       size_t length,
                       HyoutaUtils::EndianUtils::Endianness e) {
    using namespace HyoutaUtils::MemRead;
    using HyoutaUtils::EndianUtils::FromEndian;

    if (length < 8) {
        return false;
    }

    const uint32_t unknown = FromEndian(ReadUInt32(&buffer[0]), e);
    const uint32_t fileCount = FromEndian(ReadUInt32(&buffer[4]), e);
    const size_t headerSize = 8 + (static_cast<size_t>(pkg.FileCount) * 0x50);
    if (length < headerSize) {
        return false;
    }

    auto files = std::make_unique<PkgFile[]>(fileCount);
    size_t pos = 8;
    for (size_t i = 0; i < fileCount; ++i) {
        auto& f = files[i];

        f.Filename = ReadArray<0x40>(&buffer[pos]);
        f.UncompressedSize = FromEndian(ReadUInt32(&buffer[pos + 0x40]), e);
        f.CompressedSize = FromEndian(ReadUInt32(&buffer[pos + 0x44]), e);
        f.DataPosition = FromEndian(ReadUInt32(&buffer[pos + 0x48]), e);
        f.Flags = FromEndian(ReadUInt32(&buffer[pos + 0x4c]), e);

        if (f.CompressedSize < length && f.DataPosition <= length - f.CompressedSize) {
            f.Data = buffer + f.DataPosition;
        } else {
            f.Data = nullptr;
        }

        pos += 0x50;
    }

    pkg.Files = std::move(files);
    pkg.FileCount = fileCount;
    pkg.Unknown = unknown;

    return true;
}

bool CreatePkgInMemory(std::unique_ptr<char[]>& buffer,
                       size_t& bufferLength,
                       const PkgHeader& pkg,
                       HyoutaUtils::EndianUtils::Endianness e) {
    return CreatePkgInMemory(buffer, bufferLength, pkg.Files.get(), pkg.FileCount, pkg.Unknown, e);
}

bool CreatePkgInMemory(std::unique_ptr<char[]>& buffer,
                       size_t& bufferLength,
                       const PkgFile* pkgFiles,
                       uint32_t pkgFileCount,
                       uint32_t unknownValue,
                       HyoutaUtils::EndianUtils::Endianness e) {
    using namespace HyoutaUtils::MemWrite;
    using HyoutaUtils::EndianUtils::ToEndian;

    const size_t headerSize = 8 + (static_cast<size_t>(pkgFileCount) * 0x50);
    size_t totalFileSize = 0;
    for (size_t i = 0; i < pkgFileCount; ++i) {
        const auto& f = pkgFiles[i];
        if (!f.Data) {
            return false;
        }

        totalFileSize += f.CompressedSize;
    }

    const size_t dataLength = headerSize + totalFileSize;
    auto data = std::make_unique_for_overwrite<char[]>(dataLength);
    size_t currentFileSize = 0;

    WriteUInt32(&data[0], ToEndian(unknownValue, e));
    WriteUInt32(&data[4], ToEndian(pkgFileCount, e));
    size_t pos = 8;
    for (size_t i = 0; i < pkgFileCount; ++i) {
        const auto& f = pkgFiles[i];
        WriteArray(&data[pos], f.Filename);
        WriteUInt32(&data[pos + 0x40], ToEndian(f.UncompressedSize, e));
        WriteUInt32(&data[pos + 0x44], ToEndian(f.CompressedSize, e));
        WriteUInt32(&data[pos + 0x48],
                    ToEndian(static_cast<uint32_t>(headerSize + currentFileSize), e));
        WriteUInt32(&data[pos + 0x4c], ToEndian(f.Flags, e));

        pos += 0x50;
        currentFileSize += f.CompressedSize;
    }

    for (size_t i = 0; i < pkgFileCount; ++i) {
        auto& f = pkgFiles[i];
        std::memcpy(&data[pos], f.Data, f.CompressedSize);
        pos += f.CompressedSize;
    }

    buffer = std::move(data);
    bufferLength = dataLength;

    return true;
}
} // namespace SenLib
